src/core/lib/channel/connected_channel.cc

```diff
@@ -203,39 +242,606 @@ static void connected_channel_get_channel_info(
+    }
+
+   private:
+    ClientStream* impl_;
+  };
+  using StreamPtr = std::unique_ptr<grpc_stream, StreamDeleter>;
+
+  void SchedulePush() {
+    if (absl::exchange(scheduled_push_, true)) return;
+    IncrementRefCount("push");
+    ExecCtx::Run(DEBUG_LOCATION, &push_, GRPC_ERROR_NONE);
+  }
+
+  std::string ActiveOpsString() const ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    std::vector<std::string> ops;
+    if (finished_) ops.push_back("FINISHED");
+    // Pushes
+    std::vector<std::string> pushes;
+    if (push_metadata_) pushes.push_back("metadata");
+    if (push_send_message_) pushes.push_back("send_message");
+    if (push_recv_message_) pushes.push_back("recv_message");
+    if (!pushes.empty()) {
+      ops.push_back(
+          absl::StrCat(scheduled_push_ ? "push:" : "unscheduled-push:",
+                       absl::StrJoin(pushes, ",")));
+    } else if (scheduled_push_) {
+      ops.push_back("push:nothing");
+    }
+    // Results from transport
+    std::vector<std::string> queued;
+    if (queued_initial_metadata_) queued.push_back("initial_metadata");
+    if (queued_trailing_metadata_) queued.push_back("trailing_metadata");
+    if (!queued.empty()) {
+      ops.push_back(absl::StrCat("queued:", absl::StrJoin(queued, ",")));
+    }
+    // Send message
+    std::string send_message_state = SendMessageString();
+    if (send_message_state != "WAITING") {
+      ops.push_back(absl::StrCat("send_message:", send_message_state));
+    }
+    // Receive message
+    std::string recv_message_state = RecvMessageString();
+    if (recv_message_state != "IDLE") {
+      ops.push_back(absl::StrCat("recv_message:", recv_message_state));
+    }
+    return absl::StrJoin(ops, " ");
+  }
+
+  std::string SendMessageString() const ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    return Match(
+        send_message_state_, [](Idle) -> std::string { return "IDLE"; },
+        [](Closed) -> std::string { return "CLOSED"; },
+        [](const PipeReceiver<MessageHandle>::NextType&) -> std::string {
+          return "WAITING";
+        },
+        [](SendMessageToTransport) -> std::string { return "SENDING"; });
+  }
+
+  std::string RecvMessageString() const ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    return Match(
+        recv_message_state_, [](Idle) -> std::string { return "IDLE"; },
+        [](Closed) -> std::string { return "CLOSED"; },
+        [](const PendingReceiveMessage&) -> std::string { return "WAITING"; },
+        [](const absl::optional<MessageHandle>& message) -> std::string {
+          return absl::StrCat(
+              "READY:", message.has_value()
+                            ? absl::StrCat((*message)->payload()->Length(), "b")
+                            : "EOS");
+        },
+        [](const PipeSender<MessageHandle>::PushType&) -> std::string {
+          return "PUSHING";
+        });
+  }
+
+  Mutex mu_;
+  bool requested_metadata_ = false;
+  bool push_metadata_ = false;
+  bool push_send_message_ = false;
```

Comment:
I think these two need to be atomics, because they are read and written both in the activity and in `Push()`, which runs outside the activity.

Reply:
fixed